SDWebImage/SDDiskCache.h

```diff
@@ -0,0 +1,165 @@
+ */
+- (BOOL)containsDataForKey:(NSString *)key;
+
+/**
+ Returns the data associated with a given key.
+ This method may blocks the calling thread until file read finished.
+ 
+ @param key A string identifying the data. If nil, just return nil.
+ @return The value associated with key, or nil if no value is associated with key.
+ */
+- (nullable NSData *)dataForKey:(NSString *)key;
+
+/**
+ Sets the value of the specified key in the cache.
+ This method may blocks the calling thread until file write finished.
+ 
+ @param data The data to be stored in the cache.
+ @param key    The key with which to associate the value. If nil, this method has no effect.
+ */
+- (void)setData:(nullable NSData *)data forKey:(NSString *)key;
+
+/**
+ Removes the value of the specified key in the cache.
+ This method may blocks the calling thread until file delete finished.
+ 
+ @param key The key identifying the value to be removed. If nil, this method has no effect.
+ */
+- (void)removeDataForKey:(NSString *)key;
+
+/**
+ Empties the cache.
+ This method may blocks the calling thread until file delete finished.
+ */
+- (void)removeAllData;
+
+/**
+ Removes the expired data from the cache. You can choose the data to remove base on `ageLimit`, `countLimit` and `sizeLimit` options.
+ */
+- (void)removeExpiredData;
+
+/**
+ The cache path for key
+
+ @param key A string identifying the value
+ @return The cache path for key. Or nil if the key can not associate to a path
+ */
+- (nullable NSString *)cachePathForKey:(NSString *)key;
+
+/**
+ Returns the number of data in this cache.
+ This method may blocks the calling thread until file read finished.
+ 
+ @return The total data count.
+ */
+- (NSInteger)totalCount;
+
+/**
+ Returns the total size (in bytes) of data in this cache.
+ This method may blocks the calling thread until file read finished.
+ 
+ @return The total data size in bytes.
+ */
+- (NSInteger)totalSize;
+
+/**
+ The maximum expiry time of data in cache.
+ 
+ @discussion The default value is DBL_MAX, which means no limit.
+ This is not a strict limit — if data goes over the limit, the data could
+ be evicted later in background queue.
+ */
+- (NSTimeInterval)ageLimit;
+- (void)setAgeLimit:(NSTimeInterval)ageLimit;
+
+/**
+ The maximum number of data the cache should hold.
+ 
+ @discussion The default value is NSUIntegerMax, which means no limit.
+ This is not a strict limit—if the cache goes over the limit, some data in the
+ cache could be evicted later in backgound thread.
+ */
+- (NSUInteger)countLimit;
+- (void)setCountLimit:(NSUInteger)countLimit;
+
+/**
+ The maximum total size that the cache can hold before it starts evicting data.
+ 
+ @discussion The default value is NSUIntegerMax, which means no limit.
+ This is not a strict limit — if the cache goes over the limit, some data in the
+ cache could be evicted later in background queue.
+ */
+- (NSUInteger)sizeLimit;
+- (void)setSizeLimit:(NSUInteger)sizeLimit;
```

Comment:
Can we make more properties optional? With `ageLimit`, `countLimit` and `sizeLimit`, it's kind of hard for one to actually conform `SDDiskCache`.

Reply:
Well. Maybe we can totally remove this protocol design. We pass the `config` instance to the   `id<SDDiskCache>` and let they do the proper adjustment. So sometimes we add new config item, we do not need to add new method.